### ISSUE TITLE
Partial file transfer using SCP

### ISSUE THREAD

**Issue by vivek4434** [sshnet/SSH.NET#869]
Hi everyone, 

I am using SCP to to upload file to a router, I noticed that intermittently, sometimes file transfer to the device is a partial but SCP claims to be successful. when last time I faced this problem, I added a delay of 5ms in subsequent packet transfer. It worked well for  months and I started seeing this issue again. Most of the time last packet is dropped or missed. Below are some application logs:

Scp connection succeeded
Sending SCP Path without Quotes
SCP data packet send delay : 5.
Source stream length **86763**.
Total data send 16384. Last data read 16384.
Total data send 32768. Last data read 16384.
Total data send 49152. Last data read 16384.
Total data send 65536. Last data read 16384.
Total data send 81920. Last data read 16384.
Total data send 86763. Last data read 4843.
Uploading finished. Total uploaded 86763

File size on device : **83815**

Is this a known problem already?

**Comment by vivek4434** [comment]
Tagging @drieseng  - who contributed to ScpClient code.

**Comment by drieseng** [comment]
@vivek4434 I'm not aware of this issue. Can you easily reproduce this issue, and if so only with SSH.NET?
Moreover, is there any way you can help me to reproduce this issue?

**Comment by vivek4434** [comment]
@drieseng , yes the issue is easily reproducible. I tried manually transferring the file using SCP it works fine. Apart from the partial file transfer, content of the file is also corrupted. Not sure, how to repro the issue on your side. But I had a discussion with Arista team. Please check below info:

**Findings from client (us):**
When the client ends session with PSH + FIN +ACK packet the SCP server [ Arista device ] responds with an ACK and then issues FIN +ACK and gracefully closes the session but the last segments are never  received by the device. But when the client ends the session with FIN + ACK all the bytes transferred by the client machine  over scp is received at Arista end. Arista device replies the last PSH with and ACK and the FIN + ACK from client with and ACK and a sends a subsequent FIN+ACK to gracefully closes the tcp session. Wondering if it is that when the session ends with a PSH-FIN-ACK its treated as a FIN and closes the socket and never hand over the tcp segment to upper layers.

**Response from Arista team:**
It looks like the Arista device is working as expected. When the client is closing the session with PSH + FIN +ACK, this packet is getting pushed without waiting in the buffer. Hence, the Arista device is closing the session. Later, when the packets that were already in the receiving buffer (packets from x-y) are getting dropped as Arista device already received the FIN and it shall not accept anymore packets. In case it accepts, it shall be a security concern. At this stage, this shall be a half-closed TCP session.

If a TCP client is sending a FIN to the server, it means that it informs the other end that it is done with data transfer and notifies the remote end to close the Receive Buffer. Is there any reason or use case for the client to send the FIN ACK with PSH flag?


Not sure how much it will help. Let me know you thoughts.
